src/GafferScene/AttributeQuery.cpp

```diff
@@ -0,0 +1,1197 @@
+		case Gaffer::Box2fPlugTypeId:
+			switch( oid )
+			{
+				case IECore::Box2fDataTypeId:
+					result = setBox2fPlugValue(
+						*( IECore::assertedStaticCast< Gaffer::Box2fPlug const >( vplug ) ),
+						*( IECore::assertedStaticCast< Gaffer::FloatPlug >( plug ) ),
+						IECore::assertedStaticCast< IECore::Box2fData const >( object )->readable() );
+					break;
+				case IECore::Box2iDataTypeId:
+				{
+					Imath::Box2i const& b = IECore::assertedStaticCast< IECore::Box2iData const >( object )->readable();
+					result = setBox2fPlugValue(
+						*( IECore::assertedStaticCast< Gaffer::Box2fPlug const >( vplug ) ),
+						*( IECore::assertedStaticCast< Gaffer::FloatPlug >( plug ) ),
+						Imath::Box2f(
+							Imath::V2f( static_cast< float >( b.min.x ), static_cast< float >( b.min.y ) ),
+							Imath::V2f( static_cast< float >( b.max.x ), static_cast< float >( b.max.y ) ) ) );
+					break;
+				}
+				default:
+					break;
+			}
+			break;
+		case Gaffer::Box2iPlugTypeId:
+			switch( oid )
+			{
+				case IECore::Box2fDataTypeId:
+				{
+					Imath::Box2f const& b = IECore::assertedStaticCast< IECore::Box2fData const >( object )->readable();
+					result = setBox2iPlugValue(
+						*( IECore::assertedStaticCast< Gaffer::Box2iPlug const >( vplug ) ),
+						*( IECore::assertedStaticCast< Gaffer::IntPlug >( plug ) ),
+						Imath::Box2i(
+							Imath::V2i( static_cast< int >( b.min.x ), static_cast< int >( b.min.y ) ),
+							Imath::V2i( static_cast< int >( b.max.x ), static_cast< int >( b.max.y ) ) ) );
+					break;
+				}
+				case IECore::Box2iDataTypeId:
+					result = setBox2iPlugValue(
+						*( IECore::assertedStaticCast< Gaffer::Box2iPlug const >( vplug ) ),
+						*( IECore::assertedStaticCast< Gaffer::IntPlug >( plug ) ),
+						IECore::assertedStaticCast< IECore::Box2iData const >( object )->readable() );
+					break;
+				default:
+					break;
+			}
+			break;
+		default:
+			break;
+	}
+
+	return result;
+}
+
+Gaffer::Plug const* correspondingPlug( Gaffer::Plug const* const parent, Gaffer::Plug const* const child, Gaffer::Plug const* const other )
+{
+	assert( parent != 0 );
+	assert( child != 0 );
+	assert( other != 0 );
+
+	std::vector< Gaffer::Plug const* > path;
+
+	Gaffer::Plug const* plug = child;
+
+	while( plug != parent )
+	{
+		path.push_back( plug );
+		plug = plug->parent< Gaffer::Plug >();
+	}
+
+	plug = other;
+
+	while( ! path.empty() )
+	{
+		plug = plug->getChild< Gaffer::Plug >( path.back()->getName() );
+		path.pop_back();
+	}
+
+	return plug;
+}
+
+void addChildPlugsToAffectedOutputs( Gaffer::Plug const* const plug, Gaffer::DependencyNode::AffectedPlugsContainer& outputs )
+{
+	assert( plug != 0 );
+
+	if( plug->children().empty() )
+	{
+		outputs.push_back( plug );
+	}
+	else for( Gaffer::GraphComponent::ChildIterator it = plug->children().begin(), it_end = plug->children().end(); it != it_end; ++it )
+	{
+		Gaffer::Plug const* const child = IECore::runTimeCast< Gaffer::Plug const >( ( *it ).get() );
+
```

Comment:
I really would prefer not to use auto as I prefer to see the actual types involved.

Reply:
No worries - I'm a slow convert to `auto` as well, and I can see that it's useful to have the types spelled out when you're getting to grips with new code. I think `const PlugPtr &` is the type you'd need in that case then. I do think range-for and `Range` are a clear win though as they indicate the intent to visit everything - then iterator manipulations can be reserved for fiddly logic that doesn't fit that pattern, and require closer reading.

There are respected folks who argue that you should [almost always use `auto`](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/), but I haven't been able to make that leap. I think it's a clear win where the type is clearly implied already though, `auto p = make_unique<T>()` for instance. And for me I think using it in range-for was sort of the next step on from that, provided I was confident about what was in the range.

Reply:
```Gaffer::Plug::Range``` uses a boost::filter_iterator under the hood which has overhead, it returns PlugPtr by value so will result in a bunch of smart_ptr copies and associated overhead (thread sync'ing etc) which are not neccessary here. The predicate does more tests than is necessary in this case, so we would end up with 4 conditionals rather than the existing two.

Reply:
I have to confess to not having benchmarked any of this before, so thanks for bringing it up.

The first bit of good news is that we're not actually returning `PlugPtr` by value - that would indeed be bad for the reasons you mention. We're actually returning a `PlugPtr &` (as an aside, through somewhat [foul means](https://github.com/GafferHQ/gaffer/blob/master/include/Gaffer/FilteredChildIterator.h#L98-L99), and we really want to change it to [`Plug &`](https://github.com/GafferHQ/gaffer/blob/master/include/Gaffer/FilteredChildIterator.h#L67-L71) anyway).

Second bit of good news - there's actually a `Plug::OutputRange` so this can be expressed even more simply :

```else for( const auto &child : Gaffer::Plug::Range( *plug ) )
{
	addChildPlugsToAffectedOutputs( child.get(), outputs );
}
```

I do feel strongly that this is a big readability win over manual iterator fiddling and casts, and that it should be our standard idiom in Gaffer.

I did a bit of due diligence with [some benchmarking](https://github.com/johnhaddon/gaffer/commit/08d2b4a2f33a9b3e1d0ae54a8b0a53618092e893) though, and you're right that there is some overhead. From the commit in the link, I get output along these lines :

```
Raw iterators : (this PR)
 9.284458s wall, 9.260000s user + 0.000000s system = 9.260000s CPU (99.7%)
Plug Range : (my suggestion)
 11.039422s wall, 11.010000s user + 0.000000s system = 11.010000s CPU (99.7%)
Raw range-for : (mix of the two)
 9.047648s wall, 9.030000s user + 0.000000s system = 9.030000s CPU (99.8%)
Recursive iterator : (upsettingly bad  - your intuition was right there)
 20.086901s wall, 20.040000s user + 0.000000s system = 20.040000s CPU (99.8%)
```

So there is close to 20% overhead in my suggested version. But this is for a very narrow benchmark that I would be surprised has real world impact. For the improvement in readability, and expecting that we could reduce that overhead in future, the `Plug::OutputRange` is still my recommendation. For truly performance-sensitive code paths that would be unreasonable, but I don't think that's what we're looking at in this PR.

Interestingly, `range-for` with `children()` is consistently a bit faster than this PR. I don't know how, as I imagine them to be equivalent.

If we _really_ want this to be fast, there are a couple of things that do help. Dropping the `direction()` check is a mild improvement for any of the methods. But the big win is replacing `runTimeCast()` with `static_cast()`. That is safe in this case because there are checks elsewhere that prevent non-plugs being parented to plugs. And I think we have an opportunity there - `PlugRange` is constructed with the parent, so we could optimise the cast internally based on that knowledge. This the ideal to my mind : abstractions that help expressiveness and readability, but also give us a single point we can apply optimisations to with wide effect.

Reply:
> Interestingly, range-for with children() is consistently a bit faster than this PR. I don't know how, as I imagine them to be equivalent.

This turned out to be obvious in the end - range-for calls `children()` once, and this PR calls it twice. Inlining `children()` gets them back to equivalent performance, and knocks a good chunk off all the benchmarks above (almost 40% off the RecursiveIterator one). PR here : https://github.com/GafferHQ/gaffer/pull/4268.

Thanks once again for bringing up the performance side of things. Much of my focus is on trying to make an API which is easy to use and provides abstractions for common operations, and I don't always do enough to keep an eye on the performance implications. But I do think that with the right focus we can have the best of both worlds in the majority of cases.